### ISSUE TITLE
[Epic] Native GraphQL Integration

### ISSUE THREAD

**Issue by mohanarpit** [appsmithorg/appsmith#2284]
## Objective
To introduce a native GraphQL Integration on the Appsmith platform

## Requirements
1. Introduce a separate datasource called GraphQL that supports all the API authentication modes and parameters. 
2. A user should be able to add the GraphQL Query language in the post body and execute it successfully
3. We should be able to support GraphQL variables out of the box
4. We should be able to support all settings except smart substitution i.e onPageLoad, confirm before execution, timeout, encode query params
5. PRD: https://www.notion.so/appsmith/PRD-Native-GraphQL-Support-Draft-8b67d6c768a04be0a65597ecf79c5cfd

## Out of Scope
- GraphiQL Explorer
  - Provide GraphiQL's explorer like feature.
- GraphQL Specific Autocomplete
- GraphQL Subscriptions
- Query Caching
- Show variable value on hover in the evaluated value pane.
- Auto-generate variable definitions when a new variable is created in the query body.
- How to make pagination configuration easier 
  - One idea we can explore is to auto generate the cursor path based on the cursor property name. 

## Research
1. JS Substitution inside query body with quotes / w/o quotes
2. JWT Authentication
3. Relay Pagination
4. GET vs POST HTTP Methods

## RACI Matrix
| ------------- | ------------- |
| Responsible   | @sumitsum , @AmanAgarwal041 , @prapullac, @parth-appsmith  |
| Accountable   | @sumitsum |
| Consulted     | @Nikhil-Nandagopal, @mohanarpit, @ajinkyakulkarni  |
| Informed      | @nidhi-nair, @ayushpahwa |

## Other Comments / Remarks
- From Confidence: 
```
A small question on the Graphql integration
Why do we have the variables to the right VS having it at the bottom?
People may not necessarily need to use variables because we have the concept of binding. Wouldn't the space to the right be better used to display the gql schema and docs (which would actually be more valuable to users)?
```

**Comment by Johann-L-Pickard** [comment]
Hasura implements a cool version of GraphiQL which allows for easy query building and schema exploration. Being able to do this in the appsmith environment, and a way to save these queries and come back later and edit them in the same way  would be great. 

![example](https://user-images.githubusercontent.com/75929513/103076958-86654100-45d7-11eb-819f-7f5c9a0cdc81.gif)

**Comment by rberger** [comment]
Being able to create front ends for AWS AppSync (GraphQL based) and mix with other data sources would be ultra-wonderful

**Comment by thebf** [comment]
This would pair nicely with Graphile.

**Comment by benjie** [comment]
The above query builder is actually GraphiQL Explorer from OneGraph; a lot of open source projects use it (because it's awesome): https://github.com/OneGraph/graphiql-explorer

**Comment by ioloc** [comment]
> Appsmith supports raw request bodies in API calls. To call a GraphQL API, we need a POST API with a `Content-Type` header of `application/graphql`. Sample below:
> 
> <img alt="GraphQL Query" width="961" src="https://user-images.githubusercontent.com/458946/102634947-99f84e00-4178-11eb-8648-b308c74559f2.png">
> 
> While this works and gets the job done, a native integration for GraphQL will make this process much easier and simpler for users.

https://github.com/hasura/graphql-engine/issues/4830 potentially worth noting for anyone trying to use application/graphql on hasura.

**Comment by Nikhil-Nandagopal** [comment]
Next Actions
- @sumitsum @AmanAgarwal041 Get back with answers on research questions
- @ajinkyakulkarni co-ordinate with design to begin iterations
- @prapullac to come up with test plan for GraphQL
- @sumitsum to come back with timeline to pick this up

**Comment by sumitsum** [comment]
Next Actions:
- @dzhenkov Design for v1
- @AmanAgarwal041 @sumitsum Research:
  - JS Substitution inside query body with quotes / w/o quotes
    - also how the substitution will work with variables section.
   - Relay Pagination
     - Is Relay a backend or frontend library ? 
     - Will using Relay library make the experience of writing pagination queries easier for users ?
     - How to use / integrate Relay library ?
  - GET vs POST HTTP methods
    - Can both the method types be used with the same query ? 
- @sumitsum to come back with timeline to start working on GraphQL items.

Actions added to v2:
- Show variable value on hover in the evaluated value pane. 
- Provide GraphiQL's explorer like feature.

**Comment by sumitsum** [comment]
Next Actions:
- @dzhenkov - update design based on today's discussion
- @AmanAgarwal041 @sumitsum research:
  - JS Substitution inside query body with quotes / w/o quotes
    - also how the substitution will work with variables section.
   - update these research issues with findings and close them: Relay Pagination, GET vs POST
- @sumitsum - to come back with timeline to start working on GraphQL items, start thread on GraphQL action / datasource creation flow with Nikhil, check with Akash on changes to design wrt to datasource and action creation.

**Comment by sumitsum** [comment]
Meeting updates:
- @dzhenkov - update design based on last week's discussion
- @AmanAgarwal041 @sumitsum research:
  - Figure out challenges wrt JS substitution.
   - update these research issues with findings and close them: Relay Pagination, Authentication
- @AmanAgarwal041 - start looking into the FE changes required for the plugin.
- @mohanarpit @Nikhil-Nandagopal  @sumitsum - closed thread on GraphQL action / datasource creation flow + whether it will be a separate entity or part of the API body.
- @sumitsum -  start looking into the changes required on the API server side. 
- @ajinkyakulkarni - decide on whether we want to onbard the new FE engineer joining us for the GraphQL plugin

**Comment by Nikhil-Nandagopal** [comment]
@sumitsum @dzhenkov can you share the designs on this issue?

**Comment by sumitsum** [comment]
Meeting updates:
- latest figma design link: https://www.figma.com/file/ywx3aWsV1NU8s9OvoUudSz/Datasources-GraphQL?node-id=1901%3A22610 
- @ajinkyakulkarni - get a designer for the team. 
- @AmanAgarwal041 @sumitsum research:
  - Figure out challenges wrt JS substitution.
  - @sumitsum start a thread on Slack on this topic. 
- @AmanAgarwal041 - looking into the FE side changes required for the plugin.
- @mohanarpit @Nikhil-Nandagopal @sumitsum - close thread on GraphQL action / datasource creation flow + whether it will be a separate entity or part of the API body.
- @sumitsum - looking into the server side changes required for the plugin. 
- @prapullac - get a general understanding of the epic and start looking into the QA side of changes. Waiting for the new designs to be finalized and figma files to be available.

**Comment by sumitsum** [comment]
Meeting updates: 
- @rohan-arthur - get an update on whether a designer has been assigned to the team. 
- @AmanAgarwal041 @sumitsum research:
  - Figure out challenges wrt query editor / JS substitution.
  - @sumitsum start a thread on Slack on this topic. 
- @prapullac - get a general understanding of the epic and start looking into the QA side of changes. Waiting for the new designs to be finalized and figma files to be available.

Notes: 
- Based on the discussion on Slack thread we have decided that the user experience would be as follows: clicking on the GraphQL plugin icon will create a new action, like the API query, and then clicking on `save as a datasource` will create a new GraphQL datasource. We will, for now, skip the flow where datasource can be created first.
- Body and query variables should appear on the same screen, preferably side by side as in Postman.

**Comment by sumitsum** [comment]
Meeting updates: 
- @parth-appsmith  - get updated design. 
- @AmanAgarwal041 @sumitsum research:
  - Figure out challenges wrt query editor / JS substitution.
  - @sumitsum start a thread on Slack on this topic. 
- @prapullac - get a general understanding of the epic and start looking into the QA side of changes. Waiting for the new designs to be finalized and figma files to be available.
- Team - task breakdown and SP estimation once the designs have been finalized.
- @rohan-arthur share GraphQL tutorial links with the team.

**Comment by sumitsum** [comment]
Meeting updates: 
- @parth-appsmith - update designs as per today's discussion:
  - Need some way to identify that the action config page is for `GraphQL` action and not `REST API` action. Currently, both look exactly same so user might get confused about whether the action is related to `GraphQL` or `REST API`
  - Need to indicate a separation between the `Query` and `Query Variables` section in the body. 
  - Change text of `Create new API` to `Create new REST API` in the datasource creation page. 
  - @Sumit - need to figure out changes to `Pagination` tab. 
- Team - task breakdown and SP estimation once the designs have been finalized.
- @sumitsum - PoC with Spring's GraphQL framework. 
- @prapullac - get a general understanding of the epic and start looking into the QA side of changes. Waiting for the new designs to be finalized and figma files to be available.

Notes:
- Updated design links: 
  - Prototype: https://www.figma.com/proto/ywx3aWsV1NU8s9OvoUudSz/Datasources-GraphQL?page-id=1901[…]C48%2C0.31&scaling=min-zoom&starting-point-node-id=2010%3A8564
  - Figma Designs: https://www.figma.com/file/ywx3aWsV1NU8s9OvoUudSz/Datasources-GraphQL?node-id=2011%3A9307
- We would de-prioritize the design for GraphiQL explorer for now. Focus to remain on basic implementation for now. 
- Client side code changes:
  - new plugin to be created for `GraphQL` instead of adding `if - else` branches to `REST API`.
  - datasource display code needs to be re-factored to easily add / show a new plugin.

**Comment by sumitsum** [comment]
Meeting updates: 
- @parth-appsmith
  - share link for updated deisgns
  - get feedback on updated designs
    - how to make `Query Variables` section distinct ?
    - is it ok to add an icon before the variable name for all plugins ? 
    - @sumitsum - need to figure out changes to Pagination tab
  - Change text of `Create new API` to `Create new REST API` in the datasource creation page.
- @sumitsum - PoC with Spring's GraphQL framework.
- @prapullac - start working on test flow ideas. 
- **Team** - get an estimate of epic start / end timeline. 

Notes:
- Task breakdown and SP estimation: https://www.notion.so/appsmith/GraphQL-ca7c7906d0274a87a6efb726d1cedc55 (this might change a little bit over time)

**Comment by sumitsum** [comment]
Meeting updates: 
- @rohan-arthur create and share PRD (Product Requirements Document)
- @parth-appsmith
  - share link for updated deisgns
  - get feedback on updated designs
    - how to make `Query Variables` section distinct ?
    - is it ok to add an icon before the variable name for all plugins ? 
    - @sumitsum - need to figure out changes to Pagination tab
  - Change text of `Create new API` to `Create new REST API` in the datasource creation page.
- @sumitsum - PoC with Spring's GraphQL framework.
- @prapullac - started working on test flow ideas. Have written down few scenarios. 
- **Team** - get an estimate of epic start / end timeline. 

Notes:
- Task breakdown and SP estimation: https://www.notion.so/appsmith/GraphQL-ca7c7906d0274a87a6efb726d1cedc55 (this might change a little bit over time)

**Comment by sumitsum** [comment]
Meeting updates: 
* Re-estimate server side story points and timeline: https://www.notion.so/appsmith/GraphQL-ca7c7906d0274a87a6efb726d1cedc55
* Design: https://www.figma.com/file/ywx3aWsV1NU8s9OvoUudSz/Datasources-GraphQL?node-id=2010%3A8564 
* TBD: re-estimate the client side story point and timeline - @AmanAgarwal041

**Comment by sumitsum** [comment]
This week's update: 
- PoC with Spring's GraphQL framework - @sumitsum 
- start client side implementation - @AmanAgarwal041 
- review PRD shared by Rohan - @sumitsum @AmanAgarwal041

**Comment by sumitsum** [comment]
Meeting updates:
* PoC with Spring's GraphQL framework - @sumitsum
* continue client side implementation - @AmanAgarwal041
* share PRD with team for feedback - @rohan-arthur 
* share pagination implementation with team to gather feedback - @sumitsum

**Comment by sumitsum** [comment]
* close PoC with various GraphQL frameworks - @sumitsum
* continue client side implementation - @AmanAgarwal041 
  * currently working on extracting out common code from REST API plugin - currently blocked due to some conflict in code.
* share pagination implementation with team to gather feedback - @sumitsum
* add test scenarios to PRD - @rohan-arthur @prapullac 
* create / check 1-1 mapping between PRD action items and engineering document - @rohan-arthur @sumitsum @AmanAgarwal041

**Comment by sumitsum** [comment]
Meeting updates:
* create re-usable components from REST API on server side - @sumitsum 
* create editor for query body and query variable - @AmanAgarwal041 
* share pagination implementation with team to gather feedback - @sumitsum
* add test scenarios to PRD - @rohan-arthur @prapullac
* create Github issues for action items and link them to PRD - @rohan-arthur @sumitsum @AmanAgarwal041

**Comment by sumitsum** [comment]
Meeting updates:
* create re-usable components from REST API on server side - @sumitsum
* create editor for query body and query variable - @AmanAgarwal041
* share pagination implementation with team to gather feedback - @sumitsum
* add test scenarios to PRD - @rohan-arthur
* create test document - @prapullac

**Comment by sumitsum** [comment]
Meeting updates:

- create re-usable components from REST API on server side - @sumitsum
- create editor for query body and query variable - @AmanAgarwal041
- share pagination implementation with team to gather feedback - @sumitsum
- create test document - @prapullac

Note:
- @sumitsum and @AmanAgarwal041 will keep updating the PRD with status of action items.

**Comment by AmanAgarwal041** [comment]
Meeting updates:

- create re-usable components from REST API on server side - @sumitsum will be raising the PR for this by eod
- create editor for query body and query variable - @AmanAgarwal041 will be raising the PR for this by eod
- share pagination implementation with team to gather feedback - @sumitsum - Needs to be picked up
- create test document - @prapullac 

Note:
- @sumitsum and @AmanAgarwal041 will keep updating the PRD with status of action items.

**Comment by sumitsum** [comment]
Last week's updates:
* create execution flow for GraphQL on API server - @sumitsum 
* create test cases for GraphQL client code and work on pagination tab - @AmanAgarwal041 
* share pagination implementation with team to gather feedback - @sumitsum 
* create test document - @prapullac=

**Comment by sumitsum** [comment]
Meeting updates:
* create execution flow for GraphQL on API server - @sumitsum - to complete by end of week.
* create test cases for GraphQL client code and work on pagination tab - @AmanAgarwal041 - blocked on @sumitsum for server side execution flow. 
* create document on pagination issue - @rohan-arthur 
* share pagination implementation with team to gather feedback - @sumitsum 
* create test document - @prapullac

**Comment by sumitsum** [comment]
Meeting updates:
- merge execution flow for GraphQL on API server - it is currently on review - @sumitsum 
- start working on smart substitution support - @sumitsum 
- smart substitution toggle - one vs two toggles - @parth-appsmith @rohan-arthur @sumitsum 
- create test cases for GraphQL client code and work on pagination tab - @AmanAgarwal041 
- create document on pagination issue - @rohan-arthur 
- share pagination implementation with team to gather feedback - @sumitsum 
- create instance for testing by the end of next week - @sumitsum 
- start testing by the end of next week - @prapullac 
- create template for documentation - @nidhi-nair

**Comment by sumitsum** [comment]
Meeting updates:
- implement smart substitution for query variables by end of week - @sumitsum 
- start thinking about smart substitution implementation for query body - @sumitsum 
- smart substitution toggle - one vs two toggles - @parth-appsmith @rohan-arthur @sumitsum 
  - share RFC - @rohan-arthur @sumitsum 
- create test cases for GraphQL client code and work on pagination tab - @AmanAgarwal041 
- create document on pagination issue - @rohan-arthur 
- gather feedback on pagination document - @sumitsum 
- create instance for testing by the end of next week - @sumitsum 
- start testing by the end of next week - @prapullac 
- create template for documentation - @nidhi-nair

**Comment by sumitsum** [comment]
Meeting updates:
- server side implementation for cursor pagination - @sumitsum 
- design for pagination tab - @parth-appsmith 
- start thinking about smart substitution implementation for query body - @sumitsum 
- smart substitution toggle - one vs two toggles - @parth-appsmith @rohan-arthur @sumitsum 
  - share RFC - @rohan-arthur @sumitsum 
- create test cases for GraphQL client code and work on pagination tab - @AmanAgarwal041 `[In review]` 
- create template for documentation - @nidhi-nair

**Comment by sumitsum** [comment]
Last week's meeting updates: 
- server side implementation for cursor pagination - @sumitsum 
- share updated deploy preview - @sumitsum 
- updated design for pagination tab - @parth-appsmith 
- smart substitution implementation for query body - @sumitsum 
- create test cases for GraphQL client code and work on pagination tab - @AmanAgarwal041 `[In review]` 
- create template for documentation - @nidhi-nair 
- send analytics events - @sumitsum @AmanAgarwal041 
- check dropdown behavior on no variables or bad query - @AmanAgarwal041

**Comment by zzgvh** [comment]
Hello. I've begun looking at converting my apps to use GraphQL/Hasura for my backend. Currently I'm experimenting using the Authenticated API data source. It works, but has some limitations and quirks. If this epic isn't too far from fruition, I'd be very interested in being a "test subject", if that can be of any help to you.

**Comment by sumitsum** [comment]
Hi @zzgvh, thanks for showing interest in this project. Yes, it is close to completion. We can share a test instance URL with you very soon.

**Comment by sumitsum** [comment]
Hi @zzgvh you can try out the plugin [here](http://3.111.33.176/applications).

**Comment by sumitsum** [comment]
Last week's meeting updates: 
- server side implementation for cursor pagination - @sumitsum `[In review]`
- smart substitution research and implementation for query body - @sumitsum 
- refactor client datasource implementation to use same component as REST API so that OAuth can work - @AmanAgarwal041 
- create template for documentation - @nidhi-nair 
- send analytics events - @sumitsum @AmanAgarwal041

**Comment by sumitsum** [comment]
Meeting updates:
* server side implementation for cursor pagination - @sumitsum [In review]
* smart substitution research and implementation for query body - @sumitsum
* fix client side issues mentioned on deploy preview Slack thread - @AmanAgarwal041 `[In progress]`
* QA to start 2nd round of test next week onwards - @yatinappsmith @prapullac 
* create template for documentation - @nidhi-nair
* send analytics events - @sumitsum @AmanAgarwal041

**Comment by zzgvh** [comment]
> Hi @zzgvh you can try out the plugin [here](http://3.111.33.176/applications).

Hey @sumitsum, awesome! I will try to find time to test this out! 😉 BTW I see there's an appsmith-ce Docker container tagged `test-graphql`, is that running this work you're doing?

**Comment by sumitsum** [comment]
Hi @zzgvh yes. It is the same docker container that is deployed on the given URL :) .

**Comment by zzgvh** [comment]
Hello again @sumitsum! Finally I have some time to run the `test-graphql` container. However things are not 100% right, and I need some guidance to figure it out. When I spin up the `test-graphql` container everything seemingly works, and I can even log in with the identity I use for my regular development where I currently run `appsmith-ce:v1.7.5`. However upon entry there are no organisations or apps visible. And what's worse is that when I try to crate an new org I get the message "Uh oh. You do not have permissions to do: Create organization". So I am very much stuck... Any ideas?

**Comment by zzgvh** [comment]
After some fiddling I got it working. I realized I had to create a new identity, which became admin for a new org where I could create a GraphQL data source and hook it up to Hasura. Looks good!

![image](https://user-images.githubusercontent.com/142613/179187606-6a71a7a3-4dd8-43f2-9dd8-4fc8c730a766.png)

Seems like I cannot move forward with migrating to using the GraphQL datasource for now. When importing I get a complaint that my app isn't compatible, and I'm guessing it's because I'm running 1.7.5 and the GQL development branch is a little behind. But using the Autenticated API datasource works fine, even if it's a little clunkier to use parameters ;-). Keep up the good work!

**Comment by sumitsum** [comment]
Thanks @zzgvh for trying it out and providing us feedback. We are actively working to fix the remaining issues with this and ship it out as soon as possible.

**Comment by sumitsum** [comment]
Last week's updates: 
query body smart substitution implementation - @sumitsum `[In progress]`
add test cases - @AmanAgarwal041 [In progress]
QA to start 2nd round of test when updated deploy preview is available - @yatinappsmith @prapullac
create template for documentation - @nidhi-nair
send analytics events - @sumitsum @AmanAgarwal041

**Comment by sumitsum** [comment]
Updates:
- query body smart substitution implementation - @sumitsum [In progress]
- add test cases - @AmanAgarwal041 [In progress]
- QA to start 2nd round of test when updated deploy preview is available - @yatinappsmith @prapullac
- create template for documentation - @nidhi-nair
- send analytics events - @sumitsum @AmanAgarwal041

**Comment by sumitsum** [comment]
Updates:
* close all remaining server side tasks by this week - @sumitsum 
* re-order of icons - @AmanAgarwal041 
* pagination tab eval issue - @nidhi-nair @AmanAgarwal041
    - prioritize with FE coder's pod - @nidhi-nair 
* close all needs QA pipleline issues this week - @prapullac 
* find OAuth server - team 
  - if cannot find then setup an OAuth server on local - sumit
* documentation first draft - @rohan-arthur
